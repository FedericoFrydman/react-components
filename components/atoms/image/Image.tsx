import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './image.scss';

export interface ImageProps extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'css' | 'alt'> {
  /**
   * Alternate Text.
   */
  alt?: string;
  /**
   * Rounded Circle Image.
   */
  circle?: boolean;
  /**
   * Applies a color filter to the icon.
   */
  fluid?: boolean;
  /**
   * Rounded Corners.
   */
  rounded?: boolean;
  /**
   * Image Source
   */
  src: string;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLImageElement>;
}

const ImageComponent = (
  { alt = '', src, fluid = false, rounded = false, circle = false, classNames = [], onClick, ...props }: ImageProps,
  ref: ImageProps['ref'],
) => {
  const classImage = 'lsux-image';
  const cnImage = cn(
    {
      [`${classImage}`]: true,
      [`${classImage}--fluid`]: fluid,
      [`${classImage}--rounded`]: rounded,
      [`${classImage}--circle`]: circle,
      [`${classImage}--clickable`]: !!onClick,
    },
    classNames,
  );

  return (
    <img
      {...props}
      src={src}
      alt={alt ? alt + '.' : ''}
      className={cnImage}
      ref={ref}
      style={{ width: fluid ? '100%' : 'initial', ...props.style }}
      onClick={onClick}
    ></img>
  );
};

export const Image = forwardRef<HTMLImageElement, ImageProps>(ImageComponent);
