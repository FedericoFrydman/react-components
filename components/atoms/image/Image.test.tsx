import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';
import { Image, ImageProps } from './Image';

describe('Image Component', () => {
  const imageURL = `https://www.iliketowastemytime.com/sites/default/files/imagecache/blog_image/Abu-Dhabi-United-Arab-Emirates-infinity-dunes-wallpaper-by-daniel-olah-iltwmt.jpg`;
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Image src={imageURL} />);
    testId = 'test-image';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the image the provided classNames', () => {
          const image = renderImage({ classNames });

          expect(image).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLImageElement>();

    it('provides a ref to the image', () => {
      expect(renderImage({ ref })).toEqual(ref.current);
    });
  });

  const renderImage = (props: Omit<ImageProps, 'src'> = {}): HTMLElement => {
    render(<Image {...props} src={imageURL} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
