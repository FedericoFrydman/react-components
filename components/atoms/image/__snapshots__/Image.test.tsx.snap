// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component should render the component 1`] = `
<img
  alt=""
  className="lsux-image"
  src="https://www.iliketowastemytime.com/sites/default/files/imagecache/blog_image/Abu-Dhabi-United-Arab-Emirates-infinity-dunes-wallpaper-by-daniel-olah-iltwmt.jpg"
  style={
    Object {
      "width": "initial",
    }
  }
/>
`;
